shared/fonts/fonts.cpp

```diff
@@ -0,0 +1,295 @@
+    }
+  }
+
+  // FIXME: idk how GM calculates the line height 
+  // this produces similar results for me but freetype docs recommends face->size->metrics.height / 64 but thats way off compared to GM
+  rawFont.lineHeight = rawFont.yOffset = face->glyph->linearVertAdvance / 65536;
+
+  FT_Done_Face(face);
+  
+  return true;
+}
+
+unsigned char* font_pack(RawFont& font, unsigned& textureW, unsigned& textureH) {
+ 
+  if (!pack_rectangles(font.glyphs, textureW, textureH))
+    return nullptr;
+
+  // intialize blank image (the caller is responisible for deleting pxdata)
+  unsigned char* pxdata = new unsigned char[textureW * textureH];
+  for (unsigned y = 0; y < textureH; ++y) {
+    for (unsigned x = 0; x < textureW; ++x) {
+      unsigned index = y * textureW + x;
+      pxdata[index] = 0;
+    }
+  }
+
+  // now we can copy all our glyphs onto single texture
+  for (RawGlyph& rg : font.glyphs) {
+
+    GlyphMetrics glyph;
+    glyph.x = rg.horiBearingX;
+    glyph.y = -rg.horiBearingY;
+    glyph.x2 = rg.horiBearingX + rg.w();
+    glyph.y2 = -rg.horiBearingY + rg.h();
+    glyph.tx = rg.x() / double(textureW);
+    glyph.ty = rg.y() / double(textureH);
+    glyph.tx2 = (rg.x() + rg.w()) / double(textureW);
+    glyph.ty2 = (rg.y() + rg.h()) / double(textureH);
+    glyph.xs = rg.linearHoriAdvance;
+
+    // insert our glyph into the glyph range
+    rg.range->glyphs[rg.character - rg.range->start] = glyph;
+
+    // write glyph pixel data to the texture
+    for (unsigned y = 0; y < rg.h(); ++y) {
+      for (unsigned x = 0; x < rg.w(); ++x) {
+        unsigned index = (y * rg.w() + x);
+        unsigned index_out = ((y + rg.y()) * textureW + (x + rg.x()));
+        pxdata[index_out] = rg.pxdata[index];
+      }
+    }
+  }
+
+  return pxdata;
+}
+
+} // namespace fonts
+} // namespace enigma
+
+using namespace enigma::fonts;
+
+namespace enigma_user {
+
+
+#if __cplusplus > 201402L
+static inline void font_add_search_path_helper(const fs::path& p) {
+   
+  auto it = supportedExtensions.find(stringtolower(p.extension()));
+  if (it != supportedExtensions.end()) {
+    FT_Face face;
+    FT_Error error;
+    error = FT_New_Face(enigma::fonts::FontManager::GetLibrary(), p.c_str(), 0, &face );
+
+    if (error != 0)
+      DEBUG_MESSAGE("Error opening font: " + p.string(), MESSAGE_TYPE::M_ERROR);
+    
+     FT_SfntName name;
+     FT_Get_Sfnt_Name(face, TT_NAME_ID_FULL_NAME, &name);
+     std::string fontName(reinterpret_cast<char*>(name.string), name.string_len);
+     fontNames[fontName] = p;
+
+     FT_SfntName family;
+     FT_Get_Sfnt_Name(face, TT_NAME_ID_FONT_FAMILY, &family);
+     std::string fontFamily(reinterpret_cast<char*>(family.string), family.string_len);
+     fontFamilies[fontFamily] = p;
+
+     std::cout << fontFamily << std::endl;
+  }
+}
+#endif
+
+void font_add_search_path(std::string path, bool recursive) {
+  fontSearchPaths.emplace_back(path); // TODO: disallow duplicate paths?
```

Comment:
Might want to protect users from adding same path infinite times in step or something.